### PR TITLE
Add Ballerina file handling API

### PR DESCRIPTION
**Pull request by Arshardh**

### CODE REVIEW

File: modules/ballerina-native/src/main/ballerina/ballerina/lang/files/natives.bal [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+@doc:Description { value:"Closes a given file and its stream"}
+@doc:Param { value:"file: The File struct" }
+native function close (File file);
+
+@doc:Description { value:"Gets the stream from a local file"}
+@doc:Param { value:"file: The file struct" }
+native function open (File file, string accessMode);
+
+@doc:Description { value:"Get blob value of a file"}
```
Comment (sameera-jayasoma) [inline modules/ballerina-native/src/main/ballerina/ballerina/lang/files/natives.bal]: Can you please follow the same pattern when documenting functions? Please have a look at other native functions. 
here s/Get/Gets/g

File: modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/lang/files/Copy.java [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+
+            if (src.isDirectory()) {
+
+                if (!dest.exists()) {
+                    if (!dest.mkdir()) {
+                        return false;
+                    }
+                }
+
+                String files[] = src.list();
```
Comment (SupunS) [inline modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/lang/files/Copy.java]: Can't we use  `src.listFiles()` ?

File: modules/ballerina-native/src/test/java/org/ballerinalang/nativeimpl/functions/FileTest.java [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+        BStruct targetStruct = new BStruct(new StructDef(GlobalScope.getInstance()), source);
+        BValue[] args = { byteContent, targetStruct };
+
+        BLangFunctions.invoke(bLangProgram, "testWrite", args);
+        Assert.assertTrue(targetFile.exists(), "File not created");
+        Assert.assertEquals(byteContent.blobValue(), getBytesFromFile(targetFile), "Written wrong content");
+    }
+
+    @Test
+    public void testRead() throws IOException {
```
Comment (SupunS) [inline modules/ballerina-native/src/test/java/org/ballerinalang/nativeimpl/functions/FileTest.java]: Shall we add some negative test cases, and test-cases dealing with directories?

File: modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/lang/files/Copy.java (hunk 2 above)
Comment (Arshardh) [inline modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/lang/files/Copy.java]: I needed the String values of the paths to create a file object including the parent (line 101).

File: modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/lang/files/Copy.java [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+        }
+    }
+
+    private void closeQuietly(Closeable resource) {
+        try {
+            if (resource != null) {
+                resource.close();
+            }
+        } catch (Exception e) {
+            throw new BallerinaException("Exception during Resource.close()", e);
```
Comment (Arshardh) [inline modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/lang/files/Copy.java]: Fixed this by logging the error

File: modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/lang/files/Copy.java [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+        BStruct source = (BStruct) getArgument(context, 0);
+        BStruct destination = (BStruct) getArgument(context, 1);
+
+            File file = new File(source.getValue(0).stringValue());
+            File destinationFile = new File(destination.getValue(0).stringValue());
+
+            File parent = destinationFile.getParentFile();
+            if (parent != null) {
+                if (!parent.exists()) {
+                    if (!parent.mkdirs()) {
```
Comment (Arshardh) [inline modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/lang/files/Copy.java]: Fixed this

File: modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/lang/files/Close.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+        BStruct struct = (BStruct) getArgument(context, 0);
+        BufferedInputStream is = (BufferedInputStream) struct.getNativeData("inStream");
+        BufferedOutputStream os = (BufferedOutputStream) struct.getNativeData("outStream");
+        try {
+            if (is != null) {
+                is.close();
+            }
+            if (os != null) {
+                os.close();
```
Comment (Arshardh) [inline modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/lang/files/Close.java]: Fixed this by catching the exception in the closequietly method.

File: modules/ballerina-native/src/main/ballerina/ballerina/lang/files/natives.bal (hunk 1 above)
Comment (amjadhifthikar) [inline modules/ballerina-native/src/main/ballerina/ballerina/lang/files/natives.bal]: Corrected it

File: modules/ballerina-native/src/test/java/org/ballerinalang/nativeimpl/functions/FileTest.java (hunk 3 above)
Comment (Arshardh) [inline modules/ballerina-native/src/test/java/org/ballerinalang/nativeimpl/functions/FileTest.java]: Added Negative and Directory testcases

File: modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/lang/files/Open.java [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+            }
+            if (accessLC.contains("a") && accessLC.contains("w")) {
+                log.info("found both 'a' and 'w' in access mode string, opening file in append mode");
+            }
+            if (accessLC.contains("a")) {
+                createDirs(file);
+                BufferedOutputStream os = new BufferedOutputStream(new FileOutputStream(file));
+                struct.addNativeData("outStream", os);
+            } else if (accessLC.contains("w")) {
+                createDirs(file);
```
Comment (SupunS) [inline modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/lang/files/Open.java]: Need to check for `file.exists()` and `file.canWrite()`.
if the file already exists, print a warning (?)
If the file cannot be written, throw an error with a proper message.
Same before opening for reading, and appending.

File: modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/lang/files/Copy.java [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+public class Copy extends AbstractNativeFunction {
+
+    private static final Logger logger = LoggerFactory.getLogger(Copy.class);
+    @Override
+    public BValue[] execute(Context context) {
+
+        BStruct source = (BStruct) getArgument(context, 0);
+        BStruct destination = (BStruct) getArgument(context, 1);
+
+            File file = new File(source.getValue(0).stringValue());
```
Comment (SupunS) [inline modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/lang/files/Copy.java]: check for `file.exists()`

Comment (Arshardh) [inline modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/lang/files/Copy.java]: Fixed this

File: modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/lang/files/Open.java (hunk 7 above)
Comment (Arshardh) [inline modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/lang/files/Open.java]: Fixed this

Comment (Arshardh) [inline modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/lang/files/Open.java]: File not found exception is thrown once the file is not accessible. That exception is caught and a Ballerina exception is thrown subsequently.

**Pull request merged**